### PR TITLE
[2/4] CREDIT v2: CLI, Tooling & credit ask (AI assistant)

### PR DESCRIPTION
**Pull request by jsschreck**
## [2/4] CREDIT v2: CLI, Tooling & `credit ask` (AI assistant)

**Merge after**: [1/4] CREDIT v2: Trainers, Preblocks & Scheduler

---

## `credit ask` — AI assistant for weather AI

New subcommand: ask questions about your config, runs, data, or CREDIT itself using any major LLM provider.

```bash
credit ask "why is my loss diverging?"
credit ask "what does forecast_len do in the v2 trainer?"
credit ask --provider groq "summarize this config" --config conf/wxformer.yaml
```

| Provider   | Free tier | Notes                        |
|------------|-----------|------------------------------|
| Groq       | Yes       | Fast, recommended default    |
| Gemini     | Yes       | Google free API              |
| OpenAI     | No        | GPT-4o etc.                  |
| Anthropic  | No        | Agent mode (this tool)       |

Set your key: `export GROQ_API_KEY=...` (or GEMINI_API_KEY, OPENAI_API_KEY, ANTHROPIC_API_KEY).

REMEMBER NCAR EMPLOYEES, you have access to Claude. If confused email schreck@ucar.edu for an API key. 
---

### `credit/cli.py`
- Added `ask` subcommand with `--provider`, `--config`, `--model` flags
- Cleaned up subcommand registration

### Application scripts (`applications/`)
- Updated entry points to use v2 trainer and config schema

### Build env scripts (`scripts/`)
- `build_env_casper.sh` / `build_env_derecho.sh`: added `pip install --upgrade pip setuptools wheel` before CREDIT install to fix `pyproject.toml` license parse error with older setuptools
- Shared envs:
  - Casper: `/glade/work/schreck/conda-envs/credit-main-casper`
  - Derecho: `/glade/work/schreck/conda-envs/credit-main-derecho`

### CI workflow (`.github/workflows/`)
- Added lint + test steps for CLI module
- Ruff checks scoped to tracked files only

---

## Test plan
- [ ] `credit ask "hello"` with Groq free key returns a response
- [ ] `credit --help` shows `ask` subcommand
- [ ] Build env scripts complete without error on Casper and Derecho
- [ ] CI passes on this branch independently

### CODE REVIEW

**Comment by danineamati** [comment]
In case it is any inspiration for the CLI formatting, the `tyro` CLI has nice formatting (especially for `--help`). 

PyPi Link: https://pypi.org/project/tyro/

The change is that `tyro` uses data classes in python and docstrings of the class variables to automatically build the CLI help. It also allows for nested help messages.

**Comment by sgpearse** [comment]
> Would it be cleaner to pull them out into their own PR that sits between [1/4] and the CLI work?

You mention that it's extra overhead to split them, and we are also trying to move quickly.  I would suggest keeping this as-is.  

In my experience PR's will be ingested faster if they are as narrowly scoped as possible.  Just my 2c.

File: scripts/build_env_casper.sh
Comment (ggantos) [inline scripts/build_env_casper.sh]: @jsschreck - here's the error I get when I run this script and test with `credit --help`

Image

File: scripts/build_env_casper.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# Source:  /glade/u/home/schreck/.conda/envs/credit-casper
+# Target:  /glade/work/schreck/conda-envs/credit-main-casper
+#
+# The env is a full clone so the PyTorch / CUDA stack is preserved exactly.
+# miles-credit is then installed in *editable* mode from the local repo so
+# that git pulls are picked up without a reinstall.
+# ---------------------------------------------------------------------------
+
+set -euo pipefail
+
```
Comment (djgagne) [inline scripts/build_env_casper.sh:24]: I would suggest either getting rid of this script entirely or rewriting it as a bash script similar to create_derecho_env.sh.

File: scripts/build_env_derecho.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash -l
+#PBS -N credit-main-env-derecho
+#PBS -l select=1:ncpus=8:mem=64GB
+#PBS -l walltime=02:00:00
+#PBS -A NAML0001
+#PBS -q main
+#PBS -j oe
+#PBS -k eod
+
```
Comment (djgagne) [inline scripts/build_env_derecho.sh]: Get rid of this one since create_derecho_env.sh already does this generally and now has the mkl fix.

File: scripts/casper_gen2.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash -l
+#PBS -N credit_gen2
+#PBS -l select=1:ncpus=8:ngpus=1:mem=128GB:gpu_type=a100_80gb
+#PBS -l walltime=04:00:00
+#PBS -A NAML0001
+#PBS -q casper
+#PBS -j oe
+#PBS -k eod
+
```
Comment (djgagne) [inline scripts/casper_gen2.sh]: Also not general for all users. Either fix or remove.

File: credit/cli.py [hunk 4]
```diff
@@ -0,0 +1,2306 @@
...
+    """Interactive v1 → v2 config converter."""
+    import yaml
+
+    with open(args.config) as f:
+        conf = yaml.safe_load(f)
+
+    trainer_type = conf.get("trainer", {}).get("type", "unknown")
+
+    print()
+    print("=" * 62)
```
Comment (djgagne) [inline credit/cli.py]: That's a smart way to make the "=" lines show up.

Comment (djgagne) [inline credit/cli.py]: Instead of a bunch of print statements, should we use a multiline quote and call print (or logging.info) once?

File: credit/cli.py [hunk 5]
```diff
@@ -0,0 +1,2306 @@
...
+
+    is_casper = args.cluster == "casper"
+
+    r.account = _first(
+        args.account,
+        pbs_cfg.get("project") or pbs_cfg.get("account"),
+        os.environ.get("PBS_ACCOUNT"),
+        "NAML0001",
+    )
+    r.walltime = _first(args.walltime, pbs_cfg.get("walltime"), "12:00:00")
```
Comment (djgagne) [inline credit/cli.py]: Is there a way we can put all the default options here in a dictionary at the top of the file rather than hard coding them down here. It seems like they are used in a few different places, and it would be preferable to change them once instead of a bunch of times.

File: credit/cli.py [hunk 6]
```diff
@@ -0,0 +1,2306 @@
...
+            {depend_line}
+            module load ncarenv/24.12 gcc/12.4.0 ncarcompilers craype cray-mpich/8.1.29 \\
+                        cuda/12.3.2 conda/latest cudnn/9.2.0.82-12 mkl/2025.0.1
+
+            conda activate {conda_env}
+
+            REPO={repo}
+            CONFIG={config}
+
+            export PYTHONPATH="${{REPO}}:${{PYTHONPATH}}"
```
Comment (djgagne) [inline credit/cli.py]: These environment variables are all set by the conda environment activation script in create_derecho_env.sh and should not need to be set here.

File: credit/cli.py [hunk 7]
```diff
@@ -0,0 +1,2306 @@
...
+
+    If --chain was explicitly passed use it as-is.  Otherwise read
+    ``trainer.epochs`` and ``trainer.num_epoch`` from the config and compute
+    ``ceil(epochs / num_epoch)``, falling back to 1 if the keys are absent.
+    """
+    if args.chain is not None:
+        return args.chain
+    try:
+        import math
+        import yaml
```
Comment (djgagne) [inline credit/cli.py]: We could import yaml and math at the top of the file since they are common modules.

File: credit/cli.py [hunk 8]
```diff
@@ -0,0 +1,2306 @@
...
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog=textwrap.dedent("""\
+            Examples:
+              credit train -c config.yml
+              credit realtime -c config.yml --init-time 2024-01-15T00 --steps 40
+              credit rollout  -c config.yml
+              credit submit   --cluster casper  -c config.yml --gpus 1
+              credit submit   --cluster casper  -c config.yml --rollout --jobs 10
+              credit submit   --cluster derecho -c config.yml --gpus 4 --nodes 2
+              credit init     --grid 0.25deg -o my_config.yml
```
Comment (djgagne) [inline credit/cli.py]: Maybe put credit init first in the list?

File: credit/cli.py [hunk 9]
```diff
@@ -0,0 +1,2306 @@
+"""CREDIT unified command-line interface.
```
Comment (djgagne) [inline credit/cli.py]: For longer term maintenance, it would help a lot to create a cli directory and split the different cli components into separate modules in that directory.

**Review by djgagne — CHANGES_REQUESTED** [review]
See comments.

File: applications/rollout_realtime_gen2.py [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+
+from credit.datasets.era5 import ERA5Dataset
+from credit.preblock import build_preblocks, apply_preblocks
+from credit.models import load_model
+from credit.seed import seed_everything
+from credit.distributed import get_rank_info, setup, distributed_model_wrapper
+from credit.models.checkpoint import load_model_state, load_state_dict_error_handler
+from credit.output import load_metadata, make_xarray, save_netcdf_increment
+from credit.postblock import GlobalMassFixer, GlobalWaterFixer, GlobalEnergyFixer
+
```
Comment (djgagne) [inline applications/rollout_realtime_gen2.py:49]: credit.nwp methods are not being imported.

**Comment by jsschreck** [comment]
Thanks @djgagne  — addressed all comments:

1. **build_env_casper.sh** — rewritten as a general parameterized bash script (SOURCE_ENV / TARGET_ENV / REPO vars), same pattern as create_derecho_env.sh.
2. **build_env_derecho.sh** — deleted.
3. **casper_gen2.sh** — deleted.
4. **Multiple print statements** — consolidated the three job-plan blocks (train, realtime, ensemble) each into a single `print()` call with an f-string.
5. **PBS defaults dict** — `_PBS_DEFAULTS` dict now lives at the top of `_common.py` as the single source of truth for both clusters.
6. **Env vars in PBS scripts** — removed; conda activation handles them.
7. **yaml / math imports at top** — moved to module level in `_submit.py` and `_convert.py`.
8. **credit init first** — reordered in the parser.
9. **Split into cli/ package** — `credit/cli.py` (2652 lines) is gone; replaced by `credit/cli/` with `_common.py`, `_convert.py`, `_submit.py`, `_ask.py`, `_plot.py`, `_parser.py`, and `__init__.py` that re-exports everything for backward compat.
10. **rollout_realtime_gen2.py nwp import** — `from credit.nwp import build_GFS_init` is present on line 48 and imports cleanly.

**Review by charlie-becker — COMMENTED** [review]
After the implemented changes, I have successfully tested much of the API:

- `credit train`
- `credit rollout`
- `credit realtime`
- `credit submit  --cluster casper --mode train`
- `credit submit  --cluster casper --mode rollout`
- `credit submit  --cluster casper --mode realtime`

However, I did notice that submitting `credit submit  --cluster derecho` from casper still seemed to submit on casper, which I believe is incorrect.

My other general feedback would be for some slightly more robust very general examples in the default "help" (i.e. just running `credit`)

The current helper:

```
  credit init     --grid 0.25deg -o my_config.yml
  credit train    -c config.yml
  credit realtime -c config.yml --init-time 2024-01-15T00 --steps 40
  credit rollout  -c config.yml
  credit submit   --cluster casper  -c config.yml --gpus 1
  credit submit   --cluster casper  -c config.yml --mode rollout --jobs 10
  credit submit   --cluster casper  -c config.yml --mode realtime --init-time 2024-01-15T00
  credit submit   --cluster derecho -c config.yml --gpus 4 --nodes 2
```
Can you add an example for the missing functionality such as `credit plot`, `credit ask` and any other features? 

I think we are close to being able to merge -- functionality seems sufficient.

**Comment by kevinyang-cky** [comment]
@charlie-becker I think the current examples for `credit plot` are under `credit plot --h`. Would be nice if `credit -h` can include some of those.

**Review by charlie-becker — APPROVED** [review]
Good to merge!

**Review by djgagne — APPROVED** [review]
Changes look good to me.

**Comment by kevinyang-cky** [comment]
(just to document what I tested with `credit convert` and `credit plot`)

- One should be able to do `credit convert -c example-v2026.1.0.yml`(https://github.com/NCAR/miles-credit/blob/main/config/example-v2026.1.0.yml) and type in "level" for the first prompt and just hit enter for the rest of the prompts, and it will generate a converted config called example-v2026.1.0_gen2.yml. Edit the converted config to turn off the post-blocks for now, then run `credit train -c example-v2026.1.0_gen2.yml`. Should run successfully.

- DO NOT try `credit plot` with the Gen1 trained model since `cli/_plot.py` expects key name "model_state_dict", but it does not exist in the Gen1 model checkpoint.

- `Conda init` currently produces Gen1 style config file.
